routing/cross_mwm_osrm_graph.cpp

```diff
@@ -0,0 +1,263 @@
+#include "routing/cross_mwm_osrm_graph.hpp"
+#include "routing/cross_mwm_road_graph.hpp"
+#include "routing/osrm_path_segment_factory.hpp"
+
+#include "base/stl_helpers.hpp"
+
+using namespace routing;
+using namespace std;
+
+namespace
+{
+struct NodeIds
+{
+  TOsrmNodeId m_directNodeId;
+  TOsrmNodeId m_reverseNodeId;
+};
+
+OsrmMappingTypes::FtSeg GetFtSeg(Segment const & segment)
+{
+  return OsrmMappingTypes::FtSeg(segment.GetFeatureId(), segment.GetMinPointId(),
+                                 segment.GetMaxPointId());
+}
+
+/// \returns a pair of direct and reverse(backward) node id by |segment|.
+/// The first member of the pair is direct node id and the second is reverse node id.
+NodeIds GetDirectAndReverseNodeId(OsrmFtSegMapping const & segMapping, Segment const & segment)
+{
+  OsrmFtSegMapping::TFtSegVec const ftSegs = {GetFtSeg(segment)};
+  OsrmFtSegMapping::OsrmNodesT osrmNodes;
+  segMapping.GetOsrmNodes(ftSegs, osrmNodes);
+  CHECK_LESS(osrmNodes.size(), 2, ());
+  if (osrmNodes.empty())
+    return {INVALID_NODE_ID, INVALID_NODE_ID};
+
+  NodeIds const forwardNodeIds = {osrmNodes.begin()->second.first,
+                                  osrmNodes.begin()->second.second};
+  NodeIds const backwardNodeIds = {osrmNodes.begin()->second.second,
+                                   osrmNodes.begin()->second.first};
+
+  return segment.IsForward() ? forwardNodeIds : backwardNodeIds;
+}
+
+bool GetFirstValidSegment(OsrmFtSegMapping const & segMapping, NumMwmId numMwmId,
+                          TWrittenNodeId nodeId, Segment & segment)
+{
+  auto const range = segMapping.GetSegmentsRange(nodeId);
+  for (size_t segmentIndex = range.first; segmentIndex != range.second; ++segmentIndex)
+  {
+    OsrmMappingTypes::FtSeg seg;
+    // The meaning of node id in osrm is an edge between two joints.
+    // So, it's possible to consider the first valid segment from the range which returns by
+    // GetSegmentsRange().
+    segMapping.GetSegmentByIndex(segmentIndex, seg);
+
+    if (!seg.IsValid())
```

Comment:
Loop body that looks like:
```
for (...) {
  return true;
}
```
Looks quite strange. Could you please rewrite it:
```c++
for (...) {
  if (seg.IsValid()) {
    ...
    return true;
  }
}
````